programs/marginfi/src/instructions/marginfi_group/migrate_pyth_push_oracle.rs

```diff
@@ -0,0 +1,85 @@
+// Permissionless ix to migrate a Pyth-pull based bank created in 0.1.3 or earlier to the new oracle
+// model in 0.1.4, where a feed is stored directly instead of a feed id.
+
+// TODO remove in 0.1.5
+use crate::constants::{MARGINFI_SPONSORED_SHARD_ID, PYTH_PUSH_MIGRATED, PYTH_SPONSORED_SHARD_ID};
+use crate::state::marginfi_group::{Bank, MarginfiGroup};
+use crate::state::price::{OracleSetup, PythPushOraclePriceFeed};
+use crate::{live, MarginfiError, MarginfiResult};
+use anchor_lang::prelude::*;
+use pyth_solana_receiver_sdk::price_update::FeedId;
+
+pub fn migrate_pyth_push_oracle(ctx: Context<MigratePythPushOracle>) -> MarginfiResult {
+    let mut bank = ctx.accounts.bank.load_mut()?;
+
+    if bank.config.is_pyth_push_migrated() {
+        msg!("bank {:?} already migrated", &ctx.accounts.bank.key());
+        return Ok(());
+    }
+    // Staked collateral banks should use propagate instead.
+    if bank.config.oracle_setup == OracleSetup::StakedWithPythPush {
+        msg!(
+            "use propagate to migrate staked banks. Did nothing for bank: {:?}",
+            &ctx.accounts.bank.key()
+        );
+        return Ok(());
+    }
+    // All other Non-Pyth banks are unaffected, set the flag and exit, the flag does nothing for
+    // these banks.
+    if bank.config.oracle_setup != OracleSetup::PythPushOracle {
+        bank.config.update_config_flag(true, PYTH_PUSH_MIGRATED);
+        msg!(
+            "bank {:?} does not use Pyth, doing nothing",
+            &ctx.accounts.bank.key()
+        );
+        return Ok(());
+    }
+
+    let feed_id_before = bank.config.oracle_keys[0];
+    let feed_id_bytes: FeedId = bank.config.oracle_keys[0].to_bytes();
+    let (pyth_feed, _) =
+        PythPushOraclePriceFeed::find_oracle_address(PYTH_SPONSORED_SHARD_ID, &feed_id_bytes);
+    let (mrgn_feed, _) =
+        PythPushOraclePriceFeed::find_oracle_address(MARGINFI_SPONSORED_SHARD_ID, &feed_id_bytes);
+    let oracle_key = ctx.accounts.oracle.key();
+
+    // On localnet, the feed id is randomly generated, so this check is not applicable.
+    if live!() {
+        require!(
+            oracle_key == pyth_feed || oracle_key == mrgn_feed,
```

Comment:
It's probably to do with weird naming but it is confusing from the outside that we're checking the `oracle_key` is equal to the `feed address` when the purpose of the PR is migrate away from storing the feed address in the oracle keys list. This is more to do with my unfamiliarity with the pyth oracle keys + feed address system

Reply:
This is more on Pyth for calling the oracles "feeds" that have a "feed id" while we call them "oracle". The language confuses me too idk what's a better name